src/wasm-type.h

```diff
@@ -109,8 +109,13 @@ class Type {
     funcref,
     anyref,
     eqref,
+    stringref,
+    stringview_wtf8,
+    stringview_wtf16,
+    stringview_iter,
+    // From here types are non-nullable.
     i31ref,
-    dataref,
+    dataref
```

Comment:
Looks like this reordering breaks some tests in the C and JS APIs, where we care about the indexes here. So maybe this is not best. But it seems nicer to not have the nullable types split into sub-ranges... in particular it makes `isDefaultable` faster. Thoughts?

Reply:
Do people rely on the hardcoded integer values? That seems bad in the first place. I think they should only get the value by calling functions like `BinaryenFeatureAtomics()` and use it..?

Reply:
Since `stringref` and the various views are proposed to be reftypes, it almost certainly makes sense to add the new types as `HeapType::BasicHeapType` instead. I would like to remove all the reference types from `Type::BasicType` at some point as a cleanup, since using them directly (instead of as a HeapType + nullability) is almost always incorrect.

Reply:
@aheejin Yeah, I agree they should be doing that. So maybe it's fine to update the tests then.

@tlively I don't follow. This does add them to `HeapType::BasicHeapType`?

https://github.com/WebAssembly/binaryen/pull/4755/files#diff-afe568d2a9abd5777b21b5bd481afe1a51da75559098eb8feafe601be27c89c9R348-R360

Currently all ref types are added in both places.

I do like the idea to remove refs from `Type::BasicType`, but am not sure how that would work.

Reply:
Ah right. In that case I think you can remove them from here and everything should just work? It doesn't look like you've implemented the part in wasm-type.cpp where you would canonicalize a nullable reference to the BasicHeapType to use the BasicType instead, and I think that's the only thing that you would have had to remove. Basically everywhere you produce or use a `Type::stringref` you would produce or use a `Type(HeapType::string, Nullable)` instead.

Reply:
Thanks, I think I see now how we could remove all the types from here in general (though I'm not sure I see how it would work in all places, yet). For this PR, though, wouldn't it be oddly asymmetric to remove them from here? I'm just trying to match all other types in this PR.

(The original question in this thread was what to do with the constants. I buy @aheejin 's point that it's ok to just change the constants, as users shouldn't depend on them. Also it's just the constants for i31 and data, which is in the experimental GC proposal. I've updated the constants in the PR and the tests.)

Reply:
Yes, there would be some assymmetry in keeping the BasicTypes here, but if you do that, you should also update `TypeInfo::getCanonical` in wasm-type.cpp so that `Type::stringref` and `Type(HeapType::string, Nullable)` are treated as the same type. Currently they would incorrectly be distinct types. Overall I think the patch would be much simpler if you just leave out the `BasicType`s, even though that would be (temporarily) inconsistent with the other types.

Reply:
Ok, sounds good. I tried that, see the last commit (https://github.com/WebAssembly/binaryen/pull/4755/commits/bac57be4b45d75079e584b6ce6bec59ef058e79a). Definitely it makes the PR smaller, but it crashes in an infinite loop in `getCanonical / doInsert` which I'm not sure how to fix? Am I canonicalizing incorrectly there?

Reply:
So we add them only to `HeapType` and not `Type`? But we still have `dataref` or `funcref` as basic types. What's the criteria for the basic types?

Reply:
I would like to remove `dataref` and `funcref` from `BasicType` at some point as well so that we don't have two different ways to represent the same type (e.g. heap type + nullability or BasicType).

Reply:
+1, agree it would be good to use the shorthands when printing.